### ISSUE TITLE
borg mount: Permissions errors despite seemingly correct mount options

### ISSUE THREAD

**Issue by electricboogie** [borgbackup/borg#5723]
Please fill in as much of the template as possible.
-->

## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

BUG/ISSUE

#### Your borg version (borg -V).

borg 1.1.15

#### Operating system (distribution) and version.

MacOS 11.2.2
Ubuntu 20.04.2

#### How much data is handled by borg?

------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
All archives:               13.39 GB              8.54 GB              7.43 GB

                       Unique chunks         Total chunks
Chunk index:                   50058                80196

#### Full borg commandline that lead to the problem (leave away excludes and passwords)

borg mount --verbose /path/to/repo ~/mnt

## Describe the problem you're observing.

borg mount of MacOS backup mounted locally on a Linux remote repo machine mounts with the following default mount options:  "ro,nosuid,nodev,relatime,user_id=1000,group_id=1000,default_permissions" causes permissions errors like "ls: cannot open directory '.': Permission denied" and "bash: cd: ..: Permission denied."  Although user_id=1000, group_id=1000 are specified, borg seems to mount with the permissions of the MacOS machine (id=501, group=20) 

Remounting with following manual options:  "ro,nosuid,nodev,noatime,uid=1000,gid=1000,default_permissions" seems to fix this issue.

#### Can you reproduce the problem? If so, describe how. If not, describe troubleshooting steps you took before opening the issue.

Yes.

**Comment by ThomasWaldmann** [comment]
You are saying these are the **default** mount permissions.

What do you mean by that precisely? How do you know that? Provide full terminal output to show what you mean. Also show  the uid and gid of your current user.

https://www.man7.org/linux/man-pages/man8/fuse.8.html

It doesn't sound like `user_id` and `group_id` is something you shall use.

You can give `uid` and `gid` **if you want to override user and group of all files**.

Borg only parses the options it actually supports (like `uid` and `gid`) and doesn't do anything for stuff it does not support (like `user_id` and `group_id`.

**Comment by electricboogie** [comment]
I mean when I pass "borg mount" with no options these are the options with which the borgfs is mounted.  See:

borg mount --verbose path/to/repo ~/mnt
mount | grep borgfs
borgfs on ~/mnt type fuse (ro,nosuid,nodev,relatime,user_id=1000,group_id=1000,default_permissions)

I'm saying that when doing a FUSE mount it might be confusing to the user, especially when "mount" shows the borgfs mounted against what appears to be their uid/gid.  Just feels like a sharp edge to me.  Of course, there may be a good explanation, but seems very confusing to me.

**Comment by ThomasWaldmann** [comment]
your user's uid and gid when you do the mount?

ls -ld ~/mnt ?

**Comment by ThomasWaldmann** [comment]
BTW, the information in the man page is not sufficient. It just vaguely talks about `user_id` and `group_id`, but it is unclear what the effect of them actually should be.

**Comment by electricboogie** [comment]
Call me goofy but when I saw it FUSE mounted as my Linux UID/GID (1000/1000) I assumed that it was mounted that way recursively.

Default (Without UID/GID specified):

-(~/mnt:$)-> borg mount --verbose /path/to/repo ~/mnt
Mounting filesystem

-(~/mnt:$)-> ls -n ~/mnt
drwxr-xr-x 1 **1000 1000** 0 Mar  4 12:46 hostname-home-2021-03-04-18:46:31

-(~/mnt:$)-> cd ~/mnt/hostname-home-2021-03-04-18:46:31/Users/name

-(~/mnt/hostname-home-2021-03-04-18:46:31/Users/name:$)-> ls -n
drwx------ 1 **501 20** 0 Oct 13  2019  Applications
....

-(~/mnt/hostname-home-2021-03-04-18:46:31/Users/name:$)-> mount | grep borgfs
borgfs on /home/rswinford/mnt type fuse (ro,nosuid,nodev,noexec,noatime,**user_id=1000,group_id=1000**,default_permissions)

With UID/GID specified:

-(~/mnt:$)-> borg mount -o ro,nosuid,nodev,noexec,noatime,**uid=1000,gid=1000**,default_permissions --verbose /path/to/repo ~/mnt
Mounting filesystem

-(~/mnt:$)-> ls -n ~/mnt
drwxr-xr-x 1 **1000 1000** 0 Mar  4 12:46 hostname-home-2021-03-04-18:46:31

-(~/mnt/hostname-home-2021-03-04-18:46:31/Users/name:$)-> ls -n
drwx------ 1 **1000 1000** 0 Oct 13  2019  Applications

-(~/mnt:$)-> mount | grep borgfs
borgfs on /home/rswinford/mnt type fuse (ro,nosuid,nodev,noexec,noatime,**user_id=1000,group_id=1000**,default_permissions)

**Comment by electricboogie** [comment]
Yes, I see that the user_id, group_id, uid, gid mount options are all distinct at: https://www.man7.org/linux/man-pages/man8/fuse.8.html, and, yes, it seems that it is FUSE that is confusing.  I mean -- wouldn't it make sense that the mount options uid and gid to appear as mount options when I pass "borg mount -o uid=1000, gid=1000 ...; mount | grep borgfs", especially if they are distinct, but easily confused with, user_id and group_id?  

Agreed, the FUSE manpage needs some work, but perhaps we might indicate in the borg mount docs the behavior of -o uid,gid?

**Comment by ThomasWaldmann** [comment]
Looks pretty much like expected. Even the mountpoint has correct permissions.

Force-overwriting all uids/gids with the mounting user's one would be worse imho (you make useful information invisible by that).

https://github.com/borgbackup/borg/blob/1.1.15/src/borg/fuse.py#L340 guess that "popping" (getting and removing) explains why you do not see `uid` and `gid` in the mount output. Not sure what actually is expected here, instead of removing we could also let them remain there maybe (did not try that).

**Comment by electricboogie** [comment]
To me, showing the uid, gid mount options would be more intuitive.  Really I think perhaps noting this is 'borg mount' page might be helpful too, because the distinction between user_id, group_id, uid, gid thing seems confusing.  

Of course, it's up to you.  Appreciate your attention to this concern.

Perhaps something like:

"Note: A backup stores and is mounted with the permissions of the backup source.  By default, FUSE will mount only a backup's base directory with the mounting user's UID/GID (as set by the user_id and group_id mount options).  In order to view a backup's files on a non-source machine, it may be necessary to mount the backup using the extra options parameter (-o uid=,gid=) to set the mounting user's permissions recursively.  See the FUSE man page for the differences between the user_id, group_id, uid, gid mount options: https://www.man7.org/linux/man-pages/man8/fuse.8.html"

**Comment by ThomasWaldmann** [comment]
@electricboogie btw, I have just merged #5725 which might also help in some cases (same usernames/groupnames, but different uids/gids due to different system).

TODO: check/improve the docs, see previous comments.

**Comment by remyabel** [comment]
@electricboogie I created a PR. Let me know if there's anything I left out or is confusing.
